choose-index: update the docs for supporting more patterns for mv index (#16788)
pingcap/docs: choose-index.md

    |   ├─IndexRangeScan_8(Build)      | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[3,3], keep order:false, stats:pseudo                                                                                                                |
    |   ├─IndexRangeScan_9(Build)      | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[4,4], keep order:false, stats:pseudo                                                                                                                |
    |   └─TableRowIDScan_10(Probe)     | 39.94   | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo                                                                                                                             |
    +----------------------------------+---------+-----------+-----------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------+
    

>     -- json_length(j->'$.a') = 3 can't be accessed with IndexMerge directly, so TiDB can't use IndexMerge for this SQL

Reviewer: ```suggestion
    -- json_length(j->'$.a') = 3 cannot be accessed with IndexMerge directly, so TiDB cannot use IndexMerge for this SQL.
```

    | ├─IndexRangeScan_5(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[1,1], keep order:false, stats:pseudo |
    | ├─IndexRangeScan_6(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[2,2], keep order:false, stats:pseudo |
    | ├─IndexRangeScan_7(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[3,3], keep order:false, stats:pseudo |
    | ├─IndexRangeScan_8(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[4,4], keep order:false, stats:pseudo |
    | └─TableRowIDScan_9(Probe)     | 0.00    | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo              |
    +-------------------------------+---------+-----------+-----------------------------------------------------------------------------+---------------------------------------------+
    

>     -- json_length(j->'$.a') = 3 can't be accessed with IndexMerge directly, so TiDB use IndexMerge to access the other two json_contains conditions, and json_length(j->'$.a') = 3 becomes a Selection operator

Reviewer: ```suggestion
    -- json_length(j->'$.a') = 3 cannot be accessed with IndexMerge directly, so TiDB uses IndexMerge to access the other two json_contains conditions, and json_length(j->'$.a') = 3 becomes a Selection operator.
```

    | ├─IndexRangeScan_7(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[3,3], keep order:false, stats:pseudo        |
    | ├─IndexRangeScan_8(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[4,4], keep order:false, stats:pseudo        |
    | └─Selection_10(Probe)         | 0.00    | cop[tikv] |                                                                             | eq(json_length(json_extract(test.t4.j, "$.a")), 2) |
    |   └─TableRowIDScan_9          | 0.00    | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo                     |
    +-------------------------------+---------+-----------+-----------------------------------------------------------------------------+----------------------------------------------------+
    ```

- All the conditions that are used for the IndexMerge must match the semantic of `OR` or `AND` that connects them.

>     - If `json_contains` is connected with `AND`, it matches the semantic. For example:

Reviewer: ```suggestion

    - If `json_contains` is connected with `AND`, it matches the semantic. For example:
```

        | IndexMerge_9                  | 0.00    | root      |                                                                             | type: intersection                          |
        | ├─IndexRangeScan_5(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[1,1], keep order:false, stats:pseudo |
        | ├─IndexRangeScan_6(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi2(cast(json_extract(`j`, _utf8'$.b') as unsigned array)) | range:[2,2], keep order:false, stats:pseudo |
        | ├─IndexRangeScan_7(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi2(cast(json_extract(`j`, _utf8'$.b') as unsigned array)) | range:[3,3], keep order:false, stats:pseudo |
        | └─TableRowIDScan_8(Probe)     | 0.00    | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo              |
        +-------------------------------+---------+-----------+-----------------------------------------------------------------------------+---------------------------------------------+

>         -- The conditions don't match the semantic, so TiDB can't use IndexMerge for this SQL as explained above.

Reviewer: ```suggestion
        -- The conditions do not match the semantic, so TiDB cannot use IndexMerge for this SQL as explained above.
```

        |   ├─IndexRangeScan_8(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi2(cast(json_extract(`j`, _utf8'$.b') as unsigned array)) | range:[3,3], keep order:false, stats:pseudo                                                                                                             |
        |   └─TableRowIDScan_9(Probe)     | 29.97   | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo                                                                                                                          |
        +---------------------------------+---------+-----------+-----------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------+
        

>         -- The conditions don't match the semantic, so TiDB can only use IndexMerge for part of the conditions of this SQL as explained above.

Reviewer: ```suggestion
        -- The conditions do not match the semantic, so TiDB can only use IndexMerge for part of the conditions of this SQL as explained above.
```

        | ├─IndexRangeScan_6(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi2(cast(json_extract(`j`, _utf8'$.b') as unsigned array)) | range:[2,2], keep order:false, stats:pseudo |
        | ├─IndexRangeScan_7(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[3,3], keep order:false, stats:pseudo |
        | └─TableRowIDScan_8(Probe)     | 29.97   | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo              |
        +-------------------------------+---------+-----------+-----------------------------------------------------------------------------+---------------------------------------------+
        ```

    - If `json_contains` that contains multiple values is connected with `OR`, or `json_overlaps` that contains multiple values is connected with `AND`, they do not match the semantic, but they match the semantic if they only contain one value. For example:
        
        ```sql

>         -- Please refer to the above for examples of conditions that don't match the semantic. Only examples of conditions that match the semantic are provided here.

Reviewer: ```suggestion
        -- Refer to the preceding examples of conditions that do not match the semantic. Only examples of conditions that match the semantic are provided here.
```

        +-------------------------------+---------+-----------+-----------------------------------------------------------------------------+---------------------------------------------+
        | IndexMerge_8                  | 19.99   | root      |                                                                             | type: union                                 |
        | ├─IndexRangeScan_5(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[1,1], keep order:false, stats:pseudo |
        | ├─IndexRangeScan_6(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi2(cast(json_extract(`j`, _utf8'$.b') as unsigned array)) | range:[2,2], keep order:false, stats:pseudo |
        | └─TableRowIDScan_7(Probe)     | 19.99   | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo              |
        +-------------------------------+---------+-----------+-----------------------------------------------------------------------------+---------------------------------------------+
        ```

>     - If both `OR` and `AND` appear in the conditions (essentially it's nested `OR` and `AND`), the conditions that constitute the IndexMerge must all match the semantic of `OR` or the semantic of `AND`, and they can't have part of them match the semantic of `OR` while the others match the semantic of `AND`. For example:

Reviewer: ```suggestion
    - If both `OR` and `AND` appear in the conditions (nested `OR` and `AND`), the conditions that constitute the IndexMerge must all match the semantic of `OR` or the semantic of `AND`, and they cannot make part of them match the semantic of `OR` while the others match the semantic of `AND`. For example:
```

        
        ```sql
        EXPLAIN SELECT /*+ use_index_merge(t4, mvi1, mvi2) */ * FROM t4 WHERE 1 member of (j->'$.a') AND (2 member of (j->'$.b') OR 3 member of (j->'$.a'));
        EXPLAIN SELECT /*+ use_index_merge(t4, mvi1, mvi2) */ * FROM t4 WHERE 1 member of (j->'$.a') OR (2 member of (j->'$.b') AND 3 member of (j->'$.a'));
        ```

        ```sql

>         -- Only 2 member of (j->'$.b') and 3 member of (j->'$.a') that match the semantic of OR constitute the IndexMerge. 1 member of (j->'$.a') that matches the semantic of AND is not included.

Reviewer: ```suggestion
        -- Only two members of (j->'$.b') and three members of (j->'$.a') that match the semantic of OR constitute the IndexMerge. One member of (j->'$.a') that matches the semantic of AND is not included.
```

        | └─Selection_8(Probe)          | 0.00    | cop[tikv] |                                                                             | json_memberof(cast(1, json BINARY), json_extract(test.t4.j, "$.a")), or(json_memberof(cast(2, json BINARY), json_extract(test.t4.j, "$.b")), json_memberof(cast(3, json BINARY), json_extract(test.t4.j, "$.a"))) |
        |   └─TableRowIDScan_7          | 19.99   | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo                                                                                                                                                                                    |
        +-------------------------------+---------+-----------+-----------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

>         -- Only 1 member of (j->'$.a') and 3 member of (j->'$.a') that match the semantic of OR constitute the IndexMerge. 2 member of (j->'$.b') that matches the semantic of AND is not included.

Reviewer: ```suggestion
        -- Only one member of (j->'$.a') and three member of (j->'$.a') that match the semantic of OR constitute the IndexMerge. Two member of (j->'$.b') that matches the semantic of AND is not included.
```

Author: `1 member of (j->'$.a')`, `3 member of (j->'$.a')` and  `2 member of (j->'$.b')` is the expression in the SQL.

        | └─Selection_8(Probe)          | 0.02    | cop[tikv] |                                                                             | or(json_memberof(cast(1, json BINARY), json_extract(test.t4.j, "$.a")), and(json_memberof(cast(2, json BINARY), json_extract(test.t4.j, "$.b")), json_memberof(cast(3, json BINARY), json_extract(test.t4.j, "$.a")))) |
        |   └─TableRowIDScan_7          | 19.99   | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo                                                                                                                                                                                         |
        +-------------------------------+---------+-----------+-----------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
        ```

> If the conditions contain nested `OR`/`AND`, or if the conditions correspond to the index columns only after transformations such as expansion, TiDB might not be able to use IndexMerge or not be able to make full use of all conditions. It is recommended that you perform prior testing and verification for specific scenarios.

Reviewer: ```suggestion
If the conditions contain nested `OR`/`AND`, or if the conditions correspond to the index columns only after transformations such as expansion, TiDB might not be able to use IndexMerge or make full use of all conditions. It is recommended that you perform prior testing and verification for specific scenarios.
```

|   └─TableRowIDScan_7          | 0.20    | cop[tikv] | table:t6                                              | keep order:false, stats:pseudo                                                                                          |
+-------------------------------+---------+-----------+-------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
```

If several `OR`s are nested in different places in `AND`, and they need expansion to correspond to the index columns, TiDB might not be able to make full use of all conditions. For example:

```sql
EXPLAIN SELECT /*+ use_index_merge(t6, idx, idx2) */ * FROM t6 WHERE a=1 AND (1 member of (j) OR 2 member of (k)) and (b = 1 OR b = 2);
EXPLAIN SELECT /*+ use_index_merge(t6, idx, idx2) */ * FROM t6 WHERE a=1 AND ((1 member of (j) AND b = 1) OR (1 member of (j) AND b = 2) OR (2 member of (k) AND b = 1) OR (2 member of (k) AND b = 2));
```

```sql

> -- Due to current implementation limitations, (b = 1 or b = 2) doesn't constitute the IndexMerge, but become a Selection operator

Reviewer: ```suggestion
-- Due to current implementation limitations, (b = 1 or b = 2) does not constitute the IndexMerge, but becomes a Selection operator
```

| ├─IndexRangeScan_6(Build)     | 0.10    | cop[tikv] | table:t6, index:idx2(a, cast(`k` as signed array), b) | range:[1 2,1 2], keep order:false, stats:pseudo                                                                                                              |
| └─Selection_8(Probe)          | 0.20    | cop[tikv] |                                                       | eq(test.t6.a, 1), or(eq(test.t6.b, 1), eq(test.t6.b, 2)), or(json_memberof(cast(1, json BINARY), test.t6.j), json_memberof(cast(2, json BINARY), test.t6.k)) |
|   └─TableRowIDScan_7          | 0.20    | cop[tikv] | table:t6                                              | keep order:false, stats:pseudo                                                                                                                               |
+-------------------------------+---------+-----------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------+

> -- If we manually expand the two OR conditions connected with AND, TiDB can make full use of these conditions

Reviewer: ```suggestion
-- If you manually expand the two OR conditions connected with AND, TiDB can make full use of these conditions
```

| └─TableRowIDScan_6(Probe)     | 0.00    | cop[tikv] | table:t2                                                                          | keep order:false, stats:pseudo                      |
+-------------------------------+---------+-----------+-----------------------------------------------------------------------------------+-----------------------------------------------------+

> EXPLAIN SELECT /*+ use_index_merge(t2, idx) */ * FROM t2 WHERE a=1 AND JSON_CONTAINS((j->'$.path'), '[1, 2, 3]');
+-------------------------------+---------+-----------+-----------------------------------------------------------------------------------+-------------------------------------------------+
| id                            | estRows | task      | access object                                                                     | operator info                                   |
+-------------------------------+---------+-----------+-----------------------------------------------------------------------------------+-------------------------------------------------+

> | IndexMerge_9                  | 0.00    | root      |                                                                                   | type: intersection                              |

Author: This is consistent with the Chinese doc change. I just updated it with the result from the newest tidb.

| IndexMerge_9                  | 0.00    | root      |                                                                             | type: intersection                          |
| ├─IndexRangeScan_5(Build)     | 10.00   | cop[tikv] | table:t3, index:ka(a)                                                       | range:[3,3], keep order:false, stats:pseudo |
| ├─IndexRangeScan_6(Build)     | 10.00   | cop[tikv] | table:t3, index:k1(cast(json_extract(`j1`, _utf8'$.path') as signed array)) | range:[1,1], keep order:false, stats:pseudo |
| ├─IndexRangeScan_7(Build)     | 10.00   | cop[tikv] | table:t3, index:k2(cast(json_extract(`j2`, _utf8'$.path') as signed array)) | range:[2,2], keep order:false, stats:pseudo |
| └─TableRowIDScan_8(Probe)     | 0.00    | cop[tikv] | table:t3                                                                    | keep order:false, stats:pseudo              |
+-------------------------------+---------+-----------+-----------------------------------------------------------------------------+---------------------------------------------+
```

> If several `json_member_of`, `json_contains` or `json_overlaps` are connected with `OR` or `AND`, they need to meet the following requirements to access multi-valued indexes with IndexMerge:

Reviewer 2: ```suggestion
If several `json_member_of`, `json_contains` or `json_overlaps` conditions are connected with `OR` or `AND`, they need to meet the following requirements to access multi-valued indexes with IndexMerge:
```

> If several `json_member_of`, `json_contains` or `json_overlaps` are connected with `OR` or `AND`, they need to meet the following requirements to use them to access multi-valued indexes with IndexMerge:

Reviewer: ```suggestion
If several `json_member_of`, `json_contains` or `json_overlaps` are connected with `OR` or `AND`, they need to meet the following requirements to access multi-valued indexes with IndexMerge:
```

> If several `json_member_of`, `json_contains` or `json_overlaps` are connected with `OR` or `AND`, they need to meet the following requirements for TiDB to use them to access multi-valued indexes with IndexMerge:

Reviewer 3: ```suggestion
If several `json_member_of`, `json_contains` or `json_overlaps` are connected with `OR` or `AND`, they need to meet the following requirements to use them to access multi-valued indexes with IndexMerge:
```

    | Selection_5             | 8000.00  | root      |               | or(json_overlaps(json_extract(test.t4.j, "$.a"), cast("[1, 2]", json BINARY)), eq(json_length(json_extract(test.t4.j, "$.a")), 3)) |
    | └─TableReader_7         | 10000.00 | root      |               | data:TableFullScan_6                                                                                                               |
    |   └─TableFullScan_6     | 10000.00 | cop[tikv] | table:t4      | keep order:false, stats:pseudo                                                                                                     |
    +-------------------------+----------+-----------+---------------+------------------------------------------------------------------------------------------------------------------------------------+
    
    > SHOW WARNINGS;
    +---------+------+----------------------------+
    | Level   | Code | Message                    |
    +---------+------+----------------------------+
    | Warning | 1105 | IndexMerge is inapplicable |
    +---------+------+----------------------------+
    ```

> - If several conditions are connected with `AND`, some of them need to be able to be accessed with IndexMerge respectively. TiDB can only use IndexMerge to access multi-valued indexes with these conditions. For example:

Reviewer 2: ```suggestion
- For conditions connected with `AND`, some of them need to be able to be accessed with IndexMerge respectively. TiDB can only use IndexMerge to access multi-valued indexes with these conditions. For example:
```

    | ├─IndexRangeScan_6(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[2,2], keep order:false, stats:pseudo        |
    | ├─IndexRangeScan_7(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[3,3], keep order:false, stats:pseudo        |
    | ├─IndexRangeScan_8(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[4,4], keep order:false, stats:pseudo        |
    | └─Selection_10(Probe)         | 0.00    | cop[tikv] |                                                                             | eq(json_length(json_extract(test.t4.j, "$.a")), 2) |
    |   └─TableRowIDScan_9          | 0.00    | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo                     |
    +-------------------------------+---------+-----------+-----------------------------------------------------------------------------+----------------------------------------------------+
    ```

> - All the conditions that are used for the IndexMerge must match the semantic of `OR` or `AND` that connects them.

Reviewer 3: `semantic` should be replaced with `semantics` through the doc.

Reviewer: Fixed

        | IndexMerge_9                  | 0.00    | root      |                                                                             | type: intersection                          |
        | ├─IndexRangeScan_5(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[1,1], keep order:false, stats:pseudo |
        | ├─IndexRangeScan_6(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi2(cast(json_extract(`j`, _utf8'$.b') as unsigned array)) | range:[2,2], keep order:false, stats:pseudo |
        | ├─IndexRangeScan_7(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi2(cast(json_extract(`j`, _utf8'$.b') as unsigned array)) | range:[3,3], keep order:false, stats:pseudo |
        | └─TableRowIDScan_8(Probe)     | 0.00    | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo              |
        +-------------------------------+---------+-----------+-----------------------------------------------------------------------------+---------------------------------------------+

>         -- The conditions do not match the semantics, so TiDB cannot use IndexMerge for this SQL as explained above.

Reviewer 2: ```suggestion
        -- The conditions do not match the semantics, so TiDB cannot use IndexMerge for this SQL statement as explained above.
```

        |   ├─IndexRangeScan_8(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi2(cast(json_extract(`j`, _utf8'$.b') as unsigned array)) | range:[3,3], keep order:false, stats:pseudo                                                                                                             |
        |   └─TableRowIDScan_9(Probe)     | 29.97   | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo                                                                                                                          |
        +---------------------------------+---------+-----------+-----------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------+
        

>         -- The conditions do not match the semantics, so TiDB can only use IndexMerge for part of the conditions of this SQL as explained above.

Reviewer 2: ```suggestion
        -- The conditions do not match the semantics, so TiDB can only use IndexMerge for part of the conditions of this SQL statement as explained above.
```

        | IndexMerge_9                  | 29.97   | root      |                                                                             | type: union                                 |
        | ├─IndexRangeScan_5(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[1,1], keep order:false, stats:pseudo |
        | ├─IndexRangeScan_6(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi2(cast(json_extract(`j`, _utf8'$.b') as unsigned array)) | range:[2,2], keep order:false, stats:pseudo |
        | ├─IndexRangeScan_7(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[3,3], keep order:false, stats:pseudo |
        | └─TableRowIDScan_8(Probe)     | 29.97   | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo              |
        +-------------------------------+---------+-----------+-----------------------------------------------------------------------------+---------------------------------------------+
        ```

>     - If `json_contains` that contains multiple values is connected with `OR`, or `json_overlaps` that contains multiple values is connected with `AND`, they do not match the semantics, but they match the semantics if they only contain one value. For example:

Reviewer 2: ```suggestion
    - If `json_contains` conditions that contain multiple values are connected with `OR`, or `json_overlaps` conditions that contain multiple values are connected with `AND`, they do not match the semantics, but they match the semantics if they only contain one value. For example:
```

        
        ```sql

>         -- Refer to the preceding examples of conditions that do not match the semantics. Only examples of conditions that match the semantics are provided here.

Reviewer 2: ```suggestion
        -- Refer to the preceding examples for conditions that do not match the semantics. The following only provides examples of conditions that match the semantics.
```

        +-------------------------------+---------+-----------+-----------------------------------------------------------------------------+---------------------------------------------+
        | IndexMerge_8                  | 19.99   | root      |                                                                             | type: union                                 |
        | ├─IndexRangeScan_5(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi1(cast(json_extract(`j`, _utf8'$.a') as unsigned array)) | range:[1,1], keep order:false, stats:pseudo |
        | ├─IndexRangeScan_6(Build)     | 10.00   | cop[tikv] | table:t4, index:mvi2(cast(json_extract(`j`, _utf8'$.b') as unsigned array)) | range:[2,2], keep order:false, stats:pseudo |
        | └─TableRowIDScan_7(Probe)     | 19.99   | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo              |
        +-------------------------------+---------+-----------+-----------------------------------------------------------------------------+---------------------------------------------+
        ```

>     - When `OR` and `AND` occur together (essentially nested `OR` and `AND`), the conditions that make up an IndexMerge must either all match the semantics of `OR` or all match the semantics of `AND`, not partially match the semantics of `OR` and partially match the semantics of `AND`. For example:

Reviewer 2: ```suggestion
    - When both `OR` and `AND` are used to connect conditions (essentially nested `OR` and `AND`), the conditions that constitute an IndexMerge must either all match the semantics of `OR` or all match the semantics of `AND`, not partially match the semantics of `OR` and partially match the semantics of `AND`. For example:
```

>     - If both `OR` and `AND` appear in the conditions (nested `OR` and `AND`), the conditions that constitute the IndexMerge must all match the semantic of `OR` or the semantic of `AND`, and they cannot make part of them match the semantic of `OR` while the others match the semantic of `AND`. For example:

Reviewer 4: "they cannot make part of them match the semantic of `OR` while the others match the semantic" we can improve this

>     - If both `OR` and `AND` appear in the conditions (nested `OR` and `AND`), the conditions that constitute the IndexMerge must all match the semantic of `OR` or the semantic of `AND`, and they cannot make part of them match the semantic of `OR` while the others match the semantic of `AND`. For example:

Reviewer: ```suggestion
    - When `OR` and `AND` occur together (essentially nested `OR` and `AND`), the conditions that make up an IndexMerge must either all match the semantics of `OR` or all match the semantics of `AND`, not partially match the semantics of `OR` and partially match the semantics of `AND`. For example:
```

        
        ```sql
        EXPLAIN SELECT /*+ use_index_merge(t4, mvi1, mvi2) */ * FROM t4 WHERE 1 member of (j->'$.a') AND (2 member of (j->'$.b') OR 3 member of (j->'$.a'));
        EXPLAIN SELECT /*+ use_index_merge(t4, mvi1, mvi2) */ * FROM t4 WHERE 1 member of (j->'$.a') OR (2 member of (j->'$.b') AND 3 member of (j->'$.a'));
        ```

        ```sql

>         -- Only two members of (j->'$.b') and three members of (j->'$.a') that match the semantic of OR constitute the IndexMerge. One member of (j->'$.a') that matches the semantic of AND is not included.

Reviewer 4: ```suggestion
        -- Only the second member of (j->'$.b') and the third member of (j->'$.a') that match the semantic of OR constitute the IndexMerge. the first member of (j->'$.a') that matches the semantic of AND is not included.
```

Reviewer 4: from chatgpt translation？

Author: I commented above. They should be `1 member of (j->'$.a')`, `3 member of (j->'$.a')` and `2 member of (j->'$.b')` in the original SQL.
@hfxsd Should we revert to use the original content in the PR?

Reviewer: OK let's revert it.

>         -- Only two members of (j->'$.b') and three members of (j->'$.a') that match the semantic of OR constitute the IndexMerge. One member of (j->'$.a') that matches the semantic of AND is not included.

Reviewer: ```suggestion
        -- Only 2 member of (j->'$.b') and 3 members of (j->'$.a') that match the semantic of OR constitute the IndexMerge. 1 member of (j->'$.a') that matches the semantic of AND is not included.
```

        | └─Selection_8(Probe)          | 0.00    | cop[tikv] |                                                                             | json_memberof(cast(1, json BINARY), json_extract(test.t4.j, "$.a")), or(json_memberof(cast(2, json BINARY), json_extract(test.t4.j, "$.b")), json_memberof(cast(3, json BINARY), json_extract(test.t4.j, "$.a"))) |
        |   └─TableRowIDScan_7          | 19.99   | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo                                                                                                                                                                                    |
        +-------------------------------+---------+-----------+-----------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

>         -- Only one member of (j->'$.a') and three member of (j->'$.a') that match the semantic of OR constitute the IndexMerge. Two member of (j->'$.b') that matches the semantic of AND is not included.

Reviewer: ```suggestion
        -- Only 1 member of (j->'$.a') and 2 member of (j->'$.a') that match the semantic of OR constitute the IndexMerge. 2 member of (j->'$.b') that matches the semantic of AND is not included.
```

        | └─Selection_8(Probe)          | 0.02    | cop[tikv] |                                                                             | or(json_memberof(cast(1, json BINARY), json_extract(test.t4.j, "$.a")), and(json_memberof(cast(2, json BINARY), json_extract(test.t4.j, "$.b")), json_memberof(cast(3, json BINARY), json_extract(test.t4.j, "$.a")))) |
        |   └─TableRowIDScan_7          | 19.99   | cop[tikv] | table:t4                                                                    | keep order:false, stats:pseudo                                                                                                                                                                                         |
        +-------------------------------+---------+-----------+-----------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
        ```

> If the conditions contain nested `OR`/`AND`, or if the conditions correspond to only indexed columns after transformations, such as expansion, TiDB might not be able to use IndexMerge or make use of all conditions. It is recommended to verify the behavior of each specific case.

Reviewer 2: ```suggestion
If conditions contain nested `OR`/`AND`, or if conditions correspond to only indexed columns after transformations, such as expansion, TiDB might not be able to use IndexMerge or make full use of all conditions. It is recommended to verify the behavior of each specific case.
```

> If the conditions contain nested `OR`/`AND`, or if the conditions correspond to the index columns only after transformations such as expansion, TiDB might not be able to use IndexMerge or make full use of all conditions. It is recommended that you perform prior testing and verification for specific scenarios.

Reviewer 3: ```suggestion
If the conditions contain nested `OR`/`AND`, or if the conditions correspond to only indexed columns after transformations, such as expansion, TiDB might not be able to use IndexMerge or make use of all conditions. It is recommended to verify the behavior of each specific case.
```

> The following are some examples:

Reviewer 3: ```suggestion
The followings are some examples:
```

```sql
CREATE TABLE t5 (a INT, j JSON, b INT, k JSON, INDEX idx(a, (CAST(j AS SIGNED ARRAY))), INDEX idx2(b, (CAST(k as SIGNED ARRAY))));
CREATE TABLE t6 (a INT, j JSON, b INT, k JSON, INDEX idx(a, (CAST(j AS SIGNED ARRAY)), b), INDEX idx2(a, (CAST(k as SIGNED ARRAY)), b));
```

> If `AND` is nested in `OR`, and conditions connected by `AND` correspond to the exact columns of a multi-column index, TiDB can usually make full use of the conditions. For example:

Reviewer 2: ```suggestion
If `AND` is nested in conditions connected by `OR` and subconditions connected by `AND` correspond to the exact columns of a multi-column index, TiDB can usually make full use of the conditions. For example:
```

> If `AND` is nested in `OR`, and conditions connected by `AND` correspond one-to-one to the columns of a multi-column index, TiDB can usually make full use of the conditions. For example:

Reviewer 3: ```suggestion
If `AND` is nested in `OR`, and conditions connected by `AND` correspond to the exact columns of a multi-column index, TiDB can usually make full use of the conditions. For example:
```

| id                            | estRows | task      | access object                                      | operator info                                   |
+-------------------------------+---------+-----------+----------------------------------------------------+-------------------------------------------------+
| IndexMerge_8                  | 0.20    | root      |                                                    | type: union                                     |
| ├─IndexRangeScan_5(Build)     | 0.10    | cop[tikv] | table:t5, index:idx(a, cast(`j` as signed array))  | range:[1 1,1 1], keep order:false, stats:pseudo |
| ├─IndexRangeScan_6(Build)     | 0.10    | cop[tikv] | table:t5, index:idx2(b, cast(`k` as signed array)) | range:[2 2,2 2], keep order:false, stats:pseudo |
| └─TableRowIDScan_7(Probe)     | 0.20    | cop[tikv] | table:t5                                           | keep order:false, stats:pseudo                  |
+-------------------------------+---------+-----------+----------------------------------------------------+-------------------------------------------------+
```

> If a single `OR` is nested in `AND`, and it needs expansion to correspond to the indexed columns, TiDB can usually make full use of the conditions. For example:

Reviewer 2: ```suggestion
If a single `OR` is nested in conditions connected by `AND` and subconditions connected by `OR` correspond to the indexed columns after expansion, TiDB can usually make full use of the conditions. For example:
```

> If a single `OR` is nested in `AND`, and it needs expansion to correspond to the index columns, TiDB can usually make full use of the conditions. For example:

Reviewer 3: ```suggestion
If a single `OR` is nested in `AND`, and it needs expansion to correspond to the indexed columns, TiDB can usually make full use of the conditions. For example:
```

| ├─IndexRangeScan_5(Build)     | 0.10    | cop[tikv] | table:t6, index:idx(a, cast(`j` as signed array), b)  | range:[1 1,1 1], keep order:false, stats:pseudo                                                                         |
| ├─IndexRangeScan_6(Build)     | 0.10    | cop[tikv] | table:t6, index:idx2(a, cast(`k` as signed array), b) | range:[1 2,1 2], keep order:false, stats:pseudo                                                                         |
| └─Selection_8(Probe)          | 0.20    | cop[tikv] |                                                       | eq(test2.t6.a, 1), or(json_memberof(cast(1, json BINARY), test2.t6.j), json_memberof(cast(2, json BINARY), test2.t6.k)) |
|   └─TableRowIDScan_7          | 0.20    | cop[tikv] | table:t6                                              | keep order:false, stats:pseudo                                                                                          |
+-------------------------------+---------+-----------+-------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
```

> If several `OR` conditions are nested in different places in `AND`, and they need expansion to correspond to the index columns, TiDB might not be able to make full use of all conditions. For example:

Reviewer 2: ```suggestion
If multiple `OR` are nested in conditions connected by `AND` and subconditions connected by `OR` need expansion to correspond to the index columns, TiDB might not be able to make full use of all conditions. For example:
```

> If several `OR`s are nested in different places in `AND`, and they need expansion to correspond to the index columns, TiDB might not be able to make full use of all conditions. For example:

Reviewer: ```suggestion
If several `OR` conditions are nested in different places in `AND`, and they need expansion to correspond to the index columns, TiDB might not be able to make full use of all conditions. For example:
```

6 rows in set, 1 warning (0.00 sec)
```

> If several `json_member_of`, `json_contains` or `json_overlaps` are connected with `OR`/`AND`, they need to meet some requirements for TiDB to use them to access multi-valued indexes with IndexMerge:
> 1. If several conditions are connected with `OR`, each of them needs to be able to be accessed with IndexMerge respectively.
> 2. If several conditions are connected with `AND`, some of them need to be able to be accessed with IndexMerge respectively. TiDB can only use IndexMerge to access multi-valued indexes with these conditions.
> 3. All the conditions that are used for the IndexMerge must match the semantic of `OR`/`AND` that connects them. Specifically, if `json_contains` is connected with `AND`, or `json_overlaps` is connected with `OR`, or `json_member_of` is connected with `OR`/`AND`, they match the semantic; if `json_contains` that contains multiple values is connected with `OR`, or `json_overlaps` that contains multiple values is connected with `AND`, they don't match the semantic, but they match the semantic if they only contain one value.
> For example:

Reviewer: ```suggestion
If several `json_member_of`, `json_contains` or `json_overlaps` are connected with `OR`/`AND`, they need to meet the following requirements for TiDB to use them to access multi-valued indexes with IndexMerge:

- If several conditions are connected with `OR`, each of them needs to be able to be accessed with IndexMerge respectively.
- If several conditions are connected with `AND`, some of them need to be able to be accessed with IndexMerge respectively. TiDB can only use IndexMerge to access multi-valued indexes with these conditions.
- All the conditions that are used for the IndexMerge must match the semantic of `OR`/`AND` that connects them.
    - If `json_contains` is connected with `AND`, it matches the semantic.
    - If `json_overlaps` is connected with `OR`, it matches the semantic.
    - If `json_member_of` is connected with `OR`/`AND`, it matches the semantic.
    - If `json_contains` that contains multiple values is connected with `OR`, or `json_overlaps` that contains multiple values is connected with `AND`, they don't match the semantic, but they match the semantic if they only contain one value.

For example:
```

| IndexMerge_9                  | 0.00    | root      |                                                                             | type: intersection                          |
| ├─IndexRangeScan_5(Build)     | 10.00   | cop[tikv] | table:t5, index:ka(a)                                                       | range:[3,3], keep order:false, stats:pseudo |
| ├─IndexRangeScan_6(Build)     | 10.00   | cop[tikv] | table:t5, index:k1(cast(json_extract(`j1`, _utf8'$.path') as signed array)) | range:[1,1], keep order:false, stats:pseudo |
| ├─IndexRangeScan_7(Build)     | 10.00   | cop[tikv] | table:t5, index:k2(cast(json_extract(`j2`, _utf8'$.path') as signed array)) | range:[2,2], keep order:false, stats:pseudo |
| └─TableRowIDScan_8(Probe)     | 0.00    | cop[tikv] | table:t5                                                                    | keep order:false, stats:pseudo              |
+-------------------------------+---------+-----------+-----------------------------------------------------------------------------+---------------------------------------------+
```

> If the conditions contain multi-layer nested `OR`/`AND`, or if the conditions correspond to the index columns only after transformations like expansion, TiDB may not be able to use IndexMerge or not be able to make full use of all conditions. We suggest users perform prior testing and verification for specific scenarios.
> Here are examples for some scenarios:

Reviewer: ```suggestion
If the conditions contain multi-layer nested `OR`/`AND`, or if the conditions correspond to the index columns only after transformations such as expansion, TiDB might not be able to use IndexMerge or not be able to make full use of all conditions. It is recommended that you perform prior testing and verification for specific scenarios.

The following are some examples:
```

```sql
CREATE TABLE t6 (a INT, j JSON, b INT, k JSON, INDEX idx(a, (CAST(j AS SIGNED ARRAY)), b), INDEX idx2(a, (CAST(k as SIGNED ARRAY)), b));
```

> If single `OR` is nested in `AND`, and they need expansion to correspond to the index columns, TiDB can usually make full use of the conditions.

Reviewer: ```suggestion

If a single `OR` is nested in `AND`, and it needs expansion to correspond to the index columns, TiDB can usually make full use of the conditions.

```

| ├─IndexRangeScan_5(Build)     | 0.10    | cop[tikv] | table:t6, index:idx(a, cast(`j` as signed array), b)  | range:[1 1,1 1], keep order:false, stats:pseudo                                                                         |
| ├─IndexRangeScan_6(Build)     | 0.10    | cop[tikv] | table:t6, index:idx2(a, cast(`k` as signed array), b) | range:[1 2,1 2], keep order:false, stats:pseudo                                                                         |
| └─Selection_8(Probe)          | 0.20    | cop[tikv] |                                                       | eq(test2.t6.a, 1), or(json_memberof(cast(1, json BINARY), test2.t6.j), json_memberof(cast(2, json BINARY), test2.t6.k)) |
|   └─TableRowIDScan_7          | 0.20    | cop[tikv] | table:t6                                              | keep order:false, stats:pseudo                                                                                          |
+-------------------------------+---------+-----------+-------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
```

> If several `OR`s are nested in different places in `AND`, and they need expansion to correspond to the index columns, TiDB may not be able to make full use of all conditions.

Reviewer: ```suggestion
If several `OR`s are nested in different places in `AND`, and they need expansion to correspond to the index columns, TiDB might not be able to make full use of all conditions.

```

```sql

> EXPLAIN SELECT /*+ use_index_merge(t6, idx, idx2) */ * FROM t6 WHERE a=1 AND (1 member of (j) OR 2 member of (k)) and (b = 1 or b = 2);
> ```

Reviewer: ```suggestion
EXPLAIN SELECT /*+ use_index_merge(t6, idx, idx2) */ * FROM t6 WHERE a=1 AND (1 member of (j) OR 2 member of (k)) and (b = 1 or b = 2);
```

```

6 rows in set, 1 warning (0.00 sec)
```

> If several `json_member_of`, `json_contains` or `json_overlaps` are connected with `OR`/`AND`, they need to meet the following requirements for TiDB to use them to access multi-valued indexes with IndexMerge:

Reviewer: ```suggestion
If several `json_member_of`, `json_contains` or `json_overlaps` are connected with `OR` or `AND`, they need to meet the following requirements for TiDB to use them to access multi-valued indexes with IndexMerge:
```

- If several conditions are connected with `OR`, each of them needs to be able to be accessed with IndexMerge respectively.
- If several conditions are connected with `AND`, some of them need to be able to be accessed with IndexMerge respectively. TiDB can only use IndexMerge to access multi-valued indexes with these conditions.

> - All the conditions that are used for the IndexMerge must match the semantic of `OR`/`AND` that connects them.

Reviewer: ```suggestion
- All the conditions that are used for the IndexMerge must match the semantic of `OR` or `AND` that connects them.
```

    - If `json_contains` is connected with `AND`, it matches the semantic.
    - If `json_overlaps` is connected with `OR`, it matches the semantic.

>     - If `json_member_of` is connected with `OR`/`AND`, it matches the semantic.

Reviewer: ```suggestion
    - If `json_member_of` is connected with `OR` or `AND`, it matches the semantic.
```

>     - If `json_contains` that contains multiple values is connected with `OR`, or `json_overlaps` that contains multiple values is connected with `AND`, they don't match the semantic, but they match the semantic if they only contain one value.

Reviewer: ```suggestion
    - If `json_contains` that contains multiple values is connected with `OR`, or `json_overlaps` that contains multiple values is connected with `AND`, they do not match the semantic, but they match the semantic if they only contain one value.
```
